Updated defer in the dbt Cloud IDE docs (#6856)
fishtown-analytics/docs.getdbt.com: website/docs/docs/cloud/about-cloud-develop-defer.md

### Defer in the dbt Cloud IDE

For deferral in the IDE, a deploy job must have run in either the Staging environment (if it exists) or, if not, the Production environment. 

Therefore, when using defer in the IDE, it should fail if a Staging environment exists but no deploy job has run there. Metadata from Staging is unusable unless a deploy job has run successfully.

To enable defer in the dbt Cloud IDE, toggle the **Defer to production** button on the command bar. Once enabled, dbt Cloud will:

> - Pass the `--defer` flag to the command (for any command that accepts the flag).

Reviewer: ```suggestion
1. Pull down the most recent manifest from the Staging or Production environment for comparison
2. Pass the `--defer` flag to the command (for any command that accepts the flag)
```

> If you were to start developing on a new branch with [nothing in your development schema](/reference/node-selection/defer#usage), edit a single model, and run `dbt build -s state:modified` &mdash;  only the edited model would run. Any `{{ ref() }}` functions will point to the production location of the referenced models.

Reviewer: ```suggestion
If you were to start developing on a new branch with [nothing in your development schema](/reference/node-selection/defer#usage), edit a single model, and run `dbt build -s state:modified` &mdash;  only the edited model would run. Any `{{ ref() }}` functions will point to the staging or production location of the referenced models.
```

### Defer in the dbt Cloud IDE

> For deferral in the IDE, a deploy job must have run in either the Staging environment (if it exists) or, if not, the Production environment. 

Reviewer 2: ```suggestion
For deferral in the IDE, you must have production artifacts available generated by a deploy job. dbt Cloud will first check for these artifacts in your Staging environment (if it exists) or, if not, the Production environment. 
```

> For deferral in the IDE, you must have production artifacts available generated by a deploy job. dbt Cloud will first check for these artifacts in your Staging environment (if it exists) or, if not, the Production environment. 

Reviewer 3: the first sentence sounds incomplete so sugg more clarity
```suggestion
To use deferral in the IDE, you must have production artifacts generated by a deploy job. dbt Cloud will first check for these artifacts in your Staging environment (if available), or else in the Production environment. 
```

> Therefore, when using defer in the IDE, it should fail if a Staging environment exists but no deploy job has run there. Metadata from Staging is unusable unless a deploy job has run successfully.

Reviewer 2: ```suggestion
The defer feature in the IDE is not expected to work if a Staging environment exists but no deploy job has run there. The necessary metadata to power the defer feature will not exist until a deploy job has run successfully in the Staging environment.
```

> The defer feature in the IDE is not expected to work if a Staging environment exists but no deploy job has run there. The necessary metadata to power the defer feature will not exist until a deploy job has run successfully in the Staging environment.

Reviewer 3: ```suggestion
The defer feature in the IDE won't work if a Staging environment exists but no deploy job has run. This is because the necessary metadata to power defer won't exist until a deploy job has run successfully in the Staging environment.
```

> To enable defer in the dbt Cloud IDE, toggle the **Defer to production** button on the command bar. Once enabled, dbt Cloud will:

Reviewer 3: to address some of my [initial feedback on UI inconsistency](https://github.com/dbt-labs/docs.getdbt.com/pull/6856#issuecomment-2630569382)
```suggestion
To enable defer in the dbt Cloud IDE, toggle the **Defer to staging/production** button on the command bar. Once enabled, dbt Cloud will:
```

1. Pull down the most recent manifest from the Staging or Production environment for comparison
2. Pass the `--defer` flag to the command (for any command that accepts the flag)

> If you were to start developing on a new branch with [nothing in your development schema](/reference/node-selection/defer#usage), edit a single model, and run `dbt build -s state:modified` &mdash;  only the edited model would run. Any `{{ ref() }}` functions will point to the staging or production location of the referenced models.

Reviewer 3: sugg leaving the 'for example as mentioned [here](https://github.com/dbt-labs/docs.getdbt.com/pull/6856#issuecomment-2630569382) since this is an example scenario and we should let the users know that.
```suggestion
For example, if you were to start developing on a new branch with [nothing in your development schema](/reference/node-selection/defer#usage), edit a single model, and run `dbt build -s state:modified` &mdash;  only the edited model would run. Any `{{ ref() }}` functions will point to the staging or production location of the referenced models.
```

> To enable defer in the dbt Cloud IDE, toggle the **Defer to production** button on the command bar. Once enabled, dbt Cloud will:

Reviewer 3: the spacing is little off, might need to indent the paragraphs

![Screenshot 2025-01-31 at 11 24 32](https://github.com/user-attachments/assets/79c9ced8-18c6-45ea-92d3-6d1998ebe80b)

> 1. Configure deploy jobs across all three staging environments.

Reviewer 3: it might be me but are we saying users need to configure 3 staging environments? i guess, are we implying a standard practice of creating three staging environments?

> Once a Staging environment exists in a project, dbt Cloud immediately uses it to resolve cross-project references for downstream projects, without falling back to Production. This ensures that dbt Cloud consistently relies on metadata from the Staging environment, even if no successful runs have occurred in the configured Staging environment.

Reviewer 3: does this behavior apply to all defer use cases or just cross-project references? this wording makes it seem l ike this only applies to cross project ref?

> Once a Staging environment exists in a project, dbt Cloud immediately uses it to resolve cross-project references for downstream projects, without falling back to Production. This ensures that dbt Cloud consistently relies on metadata from the Staging environment, even if no successful runs have occurred in the configured Staging environment.

Reviewer 3: > This ensures that dbt Cloud consistently relies on metadata from the Staging environment, even if no successful runs have occurred in the configured Staging environment.

again, this might be me confusing things but how can dbt rely on metadata from the staging environment if no successful runs have happened? from @benmosher 's [comment here](https://dbt-labs.slack.com/archives/C01Q230T0E8/p1738250413106579?thread_ts=1738162535.339299&cid=C01Q230T0E8), it sounds like deferral in the IDE should fail if there’s no deploy job in the staging environment and that metadata from staging isn’t usable _unless_ a deploy job has run successfully.

Reviewer: it's hard for me to keep them separate but the existence of staging will impact 2 separate things:
- cross-project references: *child* projects will refer to a *parent's* staging project if it exists, else prod
- deferral: *development* environments will defer to the manifest of the latest successful run of a staging environment , if a (non-CI) job exists (even deleted!*) and has a successful run. 

in both cases, if no staging environment exists, the same behavior will occur but targeting the production environment (assuming one exists).

one tangly part is that if 
- prod environment exists and has a deploy job + a successful run
- staging environment exists but has no (non-CI) job or no successful run

the staging environment will still be referenced by both cross-project and deferral lookups. (neither will "fall back" to production)

*unclear yet whether deferral to a *deleted* job's successful run is a bug or a feature -- I need to clarify with product

Reviewer 3: confirmed here: https://github.com/dbt-labs/docs.getdbt.com/pull/6856#discussion_r1958369298
